### DISCUSSION TITLE
Dependency injection seems to waste memory

### DISCUSSION THREAD

**Discussion by tback** [fastapi/fastapi#13097] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging
from typing import Annotated
from fastapi import Depends, FastAPI


app = FastAPI()
log = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)


class A:
    def __init__(self):
        log.debug("A __init__")

    def __del__(self):
        log.debug("A __del__")


AA = Annotated[A, Depends(A)]


@app.get("/")
async def root(a: AA):
    # res = await a_function() 
    # dependency injection seems to work on routes only
    res = await a_function(a)

    return res


async def a_function(a: AA):
    log.debug(f"a_function {a}")
    return {"Hello": f"{a}"}


@app.get("/foo")
async def foo():
    return {"Hello": "Foo"}
```

### Description

I am trying to undertand the memory usage of a bigger fastapi application. I'm therefore trying to understand the dependency injection behavior and design.
* The injected object is instantiated on request as documented. However the object is deleted only when the next request using this dependency comes in. This doesn't change wether `use_cache` is `True` or `False`. Is that intentional behavior?
* Dependency injections seems to work on routes only - not on functions that are called by a route. What is the purpose of the cache parameter when the dependency has to passed around manually anyway?

### Operating System

Linux, macOS

### Operating System Details

Debian 12 on amd64
MacOS 14

### FastAPI Version

0.115.6

### Pydantic Version

2.10.3

### Python Version

Python 3.12.7

### Additional Context

Extra lines added to show beginning and end of requests. Note how `__del__` is called only when the next request is handled, but not when a request for `/foo` is handled.
```
INFO:uvicorn.error:Application startup complete.
DEBUG:server:A __init__
DEBUG:server:a_function <server.A object at 0x111755ac0>
      INFO   127.0.0.1:58513 - "GET / HTTP/1.1" 200

DEBUG:server:A __del__
DEBUG:server:A __init__
DEBUG:server:a_function <server.A object at 0x111755ac0>
      INFO   127.0.0.1:58519 - "GET / HTTP/1.1" 200

      INFO   127.0.0.1:58523 - "GET /foo HTTP/1.1" 200

DEBUG:server:A __del__
DEBUG:server:A __init__
DEBUG:server:a_function <server.A object at 0x111755ac0>
      INFO   127.0.0.1:58527 - "GET / HTTP/1.1" 200
```

**Comment by sehraramiz** [comment, marked as answer]
according to [this part of the docs](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/?h=use_cache#using-the-same-dependency-multiple-times):

> If one of your dependencies is declared multiple times for the same path operation, for example, multiple dependencies have a common sub-dependency, FastAPI will know to call that sub-dependency only once per request.
And it will save the returned value in a "cache" and pass it to all the "dependants" that need it in that specific request, instead of calling the dependency multiple times for the same request.

the dependency cache is for the time your dependencies gets multi layered and some dependencies get repeated in the hierarchy, in those cases fastapi caches a dependency when resolved and reuse it if another sub-dependency needs it (if use_cache is True)

so in your example say we have another dependency which also depend on A:

```python
import logging
from typing import Annotated
from fastapi import Depends, FastAPI


app = FastAPI()
log = logging.getLogger(__name__)
logging.basicConfig(level=logging.DEBUG)


class A:
    def __init__(self):
        log.debug("A __init__")

    def __del__(self):
        log.debug("A __del__")


AA = Annotated[A, Depends(A, use_cache=True)]


def another_dependency(a: AA) -> int:
    logging.debug(f"Another dep {a}")
    return 0


@app.get("/")
async def root(a: AA, other: Annotated[int, Depends(another_dependency)]):
    # res = await a_function()
    # dependency injection seems to work on routes only
    res = await a_function(a)

    return res


async def a_function(a: AA):
    log.debug(f"a_function {a}")
    return {"Hello": f"{a}"}


@app.get("/foo")
async def foo():
    return {"Hello": "Foo"}


```

with cache:
```
DEBUG:app:A __init__
DEBUG:root:Another dep <app.A object at 0x726db7767680>
DEBUG:app:a_function <app.A object at 0x726db7767680>
```

without cache:
```
DEBUG:app:A __init__
DEBUG:app:A __init__
DEBUG:root:Another dep <app.A object at 0x7c7e6df65280>
DEBUG:app:a_function <app.A object at 0x7c7e6df66a80>
```

so without cache we instantiated "A" twice and got two unique objects.

and about your second question, yes dependency injection in fastapi only happens in the arguments passed to route definitions and also in path operation decorators. so a_function is just a normal function with some arguments and the Depends part of it is not important until we use this function in a fastapi route or in a fastapi route path operation, then the Depends part of it becomes important.

**Comment by luzzodev** [comment]
Hi @tback i' ll try to give you some explainations.

Referring to cache behaviour, this is done only at request level. In other words it avoids to recompute the same dependecy multiple times in the same request nothing more. There is no intent or relation to cache dependency on multiple requests or in different endponts.

A valid example where it is usefull:
Imagine to have an endpoint with multiple dependencies: one to authorize an user and one to extract some data from it.
Both normally depends from a user instance that is created by fetching data from db.
Without caching dependency you should fetch twice while with cache only one.

Another example could be when you have multiple dependencies on the same route that need to db connection to do some operation.
By caching db_session you can use the same session.

To see Depends with cache in action, you can check the comment fron @sehraramiz .

About "seems to waste memory".
In your case, the dependency is solved in a separate thread, so the resolved object will be collected when the thread is released. This is managed by anyio, but in your case any call to run_in_threadpool after your response will release the old thread.

You can test it by making `async def foo()` sync. Call first `/` and then `/foo`.

About the second point:
As decribed above the the purpose of caching is only to optimize the computation, when possible for the phase of dep resolution nothing more. The caching is implemented by fastapi ( is just a simple dict ) and can be done on what fastapi has control on. ( entrypoint function of the endpoint and the depency of the route).

Hope this helps.
